File: 0216-combination-sum-iii/0216-combination-sum-iii.cpp
class Solution {
public:
    void help(vector<vector<int>> &ans, vector<int> temp, int n, int k){
        if(temp.size()==k && n==0){
            ans.push_back(temp);
            return;
        }
        for(int i= (temp.empty())?1: temp.back()+1; i<=9; i++){
            if(n-i <0) break;
            temp.push_back(i);
            help(ans, temp, n-i, k);
            temp.pop_back();
        }
        
    }
    vector<vector<int>> combinationSum3(int k, int n) {
        vector<vector<int>> ans;
        vector<int> temp;
        help(ans, temp, n, k);
        return ans;
    }
};

File: 0216-combination-sum-iii/0216-combination-sum-iii.java
class Solution {
    private void help(List<List<Integer>> ans, List<Integer> temp, int k, int n){
        if(temp.size()==k && n==0){
            List<Integer> li = new ArrayList<Integer>(temp);
            ans.add(li);
            return;
        }
        for(int i= (temp.size()==0)?1: temp.get(temp.size()-1)+1; i<=9; i++){
            if(n-i <0) break;
            temp.add(i);
            help(ans, temp, k, n-i);
            temp.remove(temp.size()-1);
        }
    }
    public List<List<Integer>> combinationSum3(int k, int n) {
        List<List<Integer>> ans= new ArrayList<>();
        List<Integer> temp = new ArrayList<>();
        help(ans, temp, k, n);
        return ans;
    }
}